pkg/services/ngalert/remote/client/mimir.go

```diff
@@ -0,0 +1,184 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+// Client contains all the methods to query the migration critical endpoints of Mimir instance, it's an interface to allow multiple implementations.
+type Client interface {
+	GetGrafanaAlertmanagerConfiguration()
+	SetGrafanaAlertmanagerConfiguration()
+	DeleteGrafanaAlertmanagerConfiguration()
+
+	GetGrafanaAlertmanagerState()
+	SetGrafanaAlertmanagerState()
+	DeleteGrafanaAlertmanagerState()
+}
+
+type Mimir struct {
+	endpoint *url.URL
+	client   http.Client
+	logger   log.Logger
+}
+
+type Config struct {
+	Address  string
+	TenantID string
+	Password string
+
+	Logger log.Logger
+}
+
+// successResponse represents a successful response from the Mimir API.
+type successResponse struct {
+	Status string `json:"status"`
+	Data   any    `json:"data"`
+}
+
+// errorResponse represents an error from the Mimir API.
+type errorResponse struct {
+	Status string `json:"status"`
+	Error1 string `json:"error"`
+	Error2 string `json:"Error"`
+}
+
+func (e *errorResponse) Error() string {
+	if e.Error1 != "" {
+		return e.Error1
+	}
+
+	return e.Error2
+}
+
+func New(cfg *Config) (*Mimir, error) {
+	endpoint, err := url.Parse(cfg.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	rt := &MimirAuthRoundTripper{
+		TenantID: cfg.TenantID,
+		Password: cfg.Password,
+		Next:     http.DefaultTransport,
+	}
+
+	c := http.Client{
+		Transport: rt,
+	}
+
+	return &Mimir{
+		endpoint: endpoint,
+		client:   c,
+		logger:   cfg.Logger,
+	}, nil
+}
+
+// do execute an HTTP requests against the specified path and method using the specified payload. It returns the HTTP response.
+func (mc *Mimir) do(ctx context.Context, p, method string, payload io.Reader, contentLength int64, out any) (*http.Response, error) {
+	pathURL, err := url.Parse(p)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := *mc.endpoint
+	endpoint.Path = path.Join(endpoint.Path, pathURL.Path)
+
+	r, err := http.NewRequestWithContext(ctx, method, endpoint.String(), payload)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Header.Set("Accept", "application/json")
+	r.Header.Set("Content-Type", "application/json")
+
+	if contentLength > 0 {
+		r.ContentLength = contentLength
+	}
+
+	logger := mc.logger.New("url", r.URL.String(), "method", r.Method)
+	resp, err := mc.client.Do(r)
+	if err != nil {
+		msg := "unable to fulfill request to the Mimir API"
+		logger.Error(msg, "err", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			logger.Error("error closing HTTP body", "err", err)
+		}
+	}()
+
+	logger = logger.New("status", resp.StatusCode)
```

Comment:
I don't think we are - if you look closely this is actually calling `.New` on an existing logger.

According to the documentation of the method it reads:

```
// New returns a new contextual Logger that has this logger's context plus the given context.
```

and looking at its implementation:

```
func (cl *ConcreteLogger) New(ctx ...any) *ConcreteLogger {
	if len(ctx) == 0 {
		root.New()
	}

	return newConcreteLogger(gokitlog.With(&cl.SwapLogger), ctx...)
}
```

It seems like we use the same logger per `&cl.SwapLogger`

Reply:
Both `gokitlog.With()` and `newConcreteLogger()` return pointers to structs, which escape to the heap once the functions exit.
Since we're calling this twice each time we call `do()`, maybe it's a good idea to just add all the context to the individual log lines:
```go
// Instead of
logger.Error(msg, "err", err)

// Do
logger.Error(msg, "err", err, "status", resp.StatusCode)
```

Reply:
Done! I've inlined all context.